components/uniset/src/builder.rs

```diff
@@ -0,0 +1,801 @@
+use std::{char, cmp, cmp::Ordering, ops::RangeBounds};
+
+use crate::{uniset::UnicodeSet, utils::deconstruct_range};
+
+/// UnicodeSet builder wrapper
+///
+/// Provides exposure to builder functions and conversion to UnicodeSet
+#[derive(Default)]
+pub struct UnicodeSetBuilder {
+    intervals: Vec<u32>,
+}
+
+impl UnicodeSetBuilder {
+    /// Returns empty UnicodeSetBuilder
+    pub fn new() -> UnicodeSetBuilder {
+        UnicodeSetBuilder { intervals: vec![] }
+    }
+
+    /// Returns a UnicodeSet and consumes the UnicodeSetBuilder
+    pub fn build(self) -> UnicodeSet {
+        UnicodeSet::from_inversion_list(self.intervals).unwrap()
+    }
+
+    /// Adds the range to the end of the builder
+    fn append(&mut self, start: u32, end: u32) {
+        self.intervals.push(start);
+        self.intervals.push(end);
+    }
+
+    /// Adds the range to the beginning of the builder in a single memory move
+    fn prepend(&mut self, start: u32, end: u32) {
+        let mut new = vec![start, end];
+        new.append(&mut self.intervals);
+        self.intervals = new;
```

Comment:
Optional:

```rust
let front = [start, end];
self.intervals.splice(0..0, &front);
```

Not sure if it's better, but it avoids allocating new memory for the vector each time.

Reply:
agreed and changed, but that syntax is not possible, has to be done as such:

```rust
let from = [start, end];
self.intervals.splice(0..0, from.iter().cloned());
```

Reply:
"cloned" is scary; use "copied" instead:

```rust
self.intervals.splice(0..0, from.iter().copied());
```

Reply:
agreed and changed

Reply:
As a side note, `cargo clippy` complains at empty ranges (`0..0` complains since technically no value gets returned from `splice`, even if we don't end up using that value regardless), but `..0` doesn't count as an empty range for some reason as well, so the syntax instead becomes `self.intervals.splice(..0, from.iter().copied());`

Reply:
That's a silly clippy error, but your workaround is fine.